cmd/prometheus/main.go

```diff
@@ -282,104 +278,182 @@ func main() {
+					case <-s2:
+						return nil
+					}
+				}
 
-	// The notifier is a dependency of the rule manager. It has to be
-	// started before and torn down afterwards.
-	go notifier.Run()
-	defer notifier.Stop()
+			},
+			func(err error) {
+				close(s2)
+			},
+		)
 
-	go ruleManager.Run()
-	defer ruleManager.Stop()
+		// Start all components while we wait for TSDB to open but only load
+		// initial config and mark ourselves as ready after it completed.
+		dbOpen := make(chan struct{})
+		s3 := make(chan struct{})
+		g.Add(
+			func() error {
+				<-dbOpen
+				if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+					level.Error(logger).Log("msg", "Error loading config", "err", err)
+					return nil
+				}
 
-	go targetManager.Run()
-	defer targetManager.Stop()
+				// Wait for reload or termination signals.
+				close(hupReady) // Unblock SIGHUP handler.
+
+				// Set web server to ready.
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive requests.")
+				<-s3
+				return nil
+			},
+			func(err error) {
+				close(s3)
+			},
+		)
 
-	// Shutting down the query engine before the rule manager will cause pending queries
-	// to be canceled and ensures a quick shutdown of the rule manager.
-	defer cancelCtx()
+		s4 := make(chan struct{})
+		g.Add(
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					level.Error(logger).Log("msg", "Opening storage failed", "err", err)
+					return nil
+				}
+				localStorage.Set(db)
+				close(dbOpen)
+				<-s4
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(s4)
+			},
+		)
 
-	errc := make(chan error)
-	go func() { errc <- webHandler.Run(ctx) }()
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
+		g.Add(
+			func() error {
+				notifier.Run()
+				return nil
+			},
+			func(err error) {
+				notifier.Stop()
+			},
+		)
 
-	<-dbOpen
+		//TODO(krasi) refactor ruleManager.Run() to be blocking to avoid using an extra blocking channel
+		s5 := make(chan struct{})
+		g.Add(
+			func() error {
+				ruleManager.Run()
+				<-s5
+				return nil
+			},
+			func(err error) {
+				ruleManager.Stop()
+				close(s5)
+			},
+		)
+		//TODO(krasi) refactor targetManager.Run() to be blocking to avoid using an extra blocking channel
+		s6 := make(chan struct{})
+		g.Add(
+			func() error {
+				targetManager.Run()
+				<-s6
+				return nil
+			},
+			func(err error) {
+				targetManager.Stop()
+				close(s6)
+			},
+		)
+		g.Add(
+			func() error {
+				if err := webHandler.Run(ctx); err != nil {
+					level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				}
+				return nil
```

Comment:
Why is nil being returned here and not the error?

Reply:
I tried to return and print the error but you loose the extra `msg` string
`"msg", "Error starting web server, exiting gracefully",`

Reply:
Exiting gracefully is expected, I don't think that needs to be logged. You could just wrap the error message `return fmt.Errorf("Error starting web server: %s", err)`.